{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter a nucleotide sequence:AAGTCGGGCTTC\n",
      "Enter a nucleotide sequence:AATATTTTCCTTCTCT\n",
      "Enter a nucleotide sequence:PWSKKLKJJJKKLZ\n",
      "Enter a nucleotide sequence:\n",
      "{'AAG': 1, 'AGT': 1, 'GTC': 1, 'TCG': 1, 'CGG': 1, 'GGG': 1, 'GGC': 1, 'GCT': 1, 'CTT': 2, 'TTC': 3, 'AAT': 1, 'ATA': 1, 'TAT': 1, 'ATT': 1, 'TTT': 2, 'TCC': 1, 'CCT': 1, 'TCT': 2, 'CTC': 1}\n",
      "{'AAGT': 1, 'AGTC': 1, 'GTCG': 1, 'TCGG': 1, 'CGGG': 1, 'GGGC': 1, 'GGCT': 1, 'GCTT': 1, 'CTTC': 2, 'AATA': 1, 'ATAT': 1, 'TATT': 1, 'ATTT': 1, 'TTTT': 1, 'TTTC': 1, 'TTCC': 1, 'TCCT': 1, 'CCTT': 1, 'TTCT': 1, 'TCTC': 1, 'CTCT': 1}\n",
      "{'AAGTC': 1, 'AGTCG': 1, 'GTCGG': 1, 'TCGGG': 1, 'CGGGC': 1, 'GGGCT': 1, 'GGCTT': 1, 'GCTTC': 1, 'AATAT': 1, 'ATATT': 1, 'TATTT': 1, 'ATTTT': 1, 'TTTTC': 1, 'TTTCC': 1, 'TTCCT': 1, 'TCCTT': 1, 'CCTTC': 1, 'CTTCT': 1, 'TTCTC': 1, 'TCTCT': 1}\n"
     ]
    }
   ],
   "source": [
    "#! /usr/bin/env python3\n",
    "class Sequence:\n",
    "    valid_dna = \"ATCG\" # creates a class attribute, which is called in sequence_type()\n",
    "    \n",
    "    def get_lookup():\n",
    "    # Obtained from https://stackoverflow.com/questions/19521905/translation-dna-to-protein\n",
    "        return {\n",
    "        'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "        'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "        'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "        'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',\n",
    "        'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "        'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "        'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "        'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "        'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "        'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "        'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "        'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "        'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "        'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "        'TAC':'Y', 'TAT':'Y', 'TAA':'*', 'TAG':'*',\n",
    "        'TGC':'C', 'TGT':'C', 'TGA':'*', 'TGG':'W',\n",
    "        }\n",
    "    \n",
    "    # creates a helper class method which detects whether a given \"seq\" is a \n",
    "    # nucleotide or protein sequence\n",
    "    def sequence_type(seq):\n",
    "        valid_seq = \"ATCG\"\n",
    "        for i in seq:\n",
    "            if i not in valid_seq:\n",
    "                return False\n",
    "        return True\n",
    "    \n",
    "    # returns reverse compliment to nucleotide sequence        \n",
    "    def rev_comp(self):\n",
    "        ret_seq = \"\"\n",
    "\n",
    "        for nuc in self.seq[::-1]:\n",
    "            if nuc == 'A':\n",
    "                ret_seq += 'T'\n",
    "            if nuc == 'C':\n",
    "                ret_seq += 'G'\n",
    "            if nuc == 'G':\n",
    "                ret_seq += 'C'\n",
    "            if nuc == 'T':\n",
    "                ret_seq += 'A'\n",
    "\n",
    "        return ret_seq\n",
    "   \n",
    "    # translates a given nucleotide sequence into all possible amino acids\n",
    "    # in all six reading frames \n",
    "    def translateSeq(self):\n",
    "        ret_protein = []\n",
    "        lookup = Sequence.get_lookup()\n",
    "        \n",
    "        # Start on the requested frame, and then skip a codons length each time\n",
    "        for frame in range(6):\n",
    "            strand = self.seq if frame < 3 else self.rev_comp()\n",
    "            trans_seq = ''\n",
    "            \n",
    "            for pos in range(frame % 3, len(strand), 3):\n",
    "                if frame < 3:    \n",
    "                    codon = strand[pos:pos+3]\n",
    "                    if codon in lookup:\n",
    "                        trans_seq += lookup[codon]\n",
    "                else:\n",
    "                    rev_codon = strand[pos:pos+3]\n",
    "                    if rev_codon in lookup:  \n",
    "                        trans_seq += lookup[rev_codon]\n",
    "            \n",
    "            ret_protein.append(trans_seq)\n",
    "    \n",
    "   \n",
    "        # if protein sequence, return sequence. If nucleotide sequence, return translation on 6 frames\n",
    "        if self.seq_type == \"P\":\n",
    "            return self.seq\n",
    "        else:\n",
    "            return ret_protein\n",
    "    \n",
    "    # instance method that returns GC content if seq_type is \"N\", else return 0\n",
    "    def gc_content(self):\n",
    "        gc_count = self.seq.count(\"G\") + self.seq.count(\"C\")\n",
    "        if self.seq_type == \"N\":\n",
    "            return float(gc_count / len(self.seq))\n",
    "        return 0\n",
    "    \n",
    "    # overload len operator to return the length of Sequence() instance\n",
    "    def __len__(self):\n",
    "        return len(self.seq)\n",
    "        \n",
    "    # concatenates sequences of the same type, else creates empty sequence of type \"N\"\n",
    "    def __add__(self, other):\n",
    "        while isinstance(self, Sequence) and isinstance (other, Sequence):\n",
    "            if self.seq_type == other.seq_type:\n",
    "                return self.seq + other.seq\n",
    "            else:\n",
    "                return Sequence()\n",
    "        return Sequence()\n",
    "      \n",
    "\n",
    "    # overloads the equality operator, ensures both objects are Sequence instances before comparing them\n",
    "    def __eq__(a, b):\n",
    "        if (not isinstance(a, Sequence)) or (not isinstance(b, Sequence)):\n",
    "            return False\n",
    "       \n",
    "        while isinstance(a, Sequence) and isinstance(b, Sequence):\n",
    "            for i in a.translateSeq():\n",
    "                if i in b.translateSeq():\n",
    "                    return True\n",
    "                \n",
    "                return False\n",
    "    \n",
    "    # ovverides __str__ method to return \"<seq_type>:<seq>\"       \n",
    "    def __str__(self):\n",
    "        return self.seq\n",
    "        \n",
    "  # init method takes two optional parameters, seq and seq_type, which have default values.\n",
    "    def __init__(self, seq =\"\", seq_type=\"N\"):\n",
    "        self.seq = seq\n",
    "        self.seq_type = seq_type\n",
    "        \n",
    "        # if sequence is provided, but no type, checks sequence type with helper method\n",
    "        if seq != \"\" and seq_type == \"\":\n",
    "            self.seq_type = \"P\" if Sequence.sequence_type(seq) == False else \"N\" \n",
    "       \n",
    "        # if sequence and type are provided, and nucleotide type is specified, checks sequence\n",
    "        # type with helper method. This allows user to override and specify a protein type for a\n",
    "        # string containing only \"AGTC\"\n",
    "        if seq != \"\" and seq_type != \"\":\n",
    "            if seq_type == \"N\":\n",
    "                self.seq_type = \"P\" if Sequence.sequence_type(seq) == False else \"N\"\n",
    "\n",
    "# Function continually prompts user for a nucleotide string, until empty string. Creates a Sequence instance\n",
    "# for each entry, and returns list of all Sequence instances.\n",
    "def sequenceList():\n",
    "    inst_list = list()\n",
    "    while True:\n",
    "        user = input(\"Enter a nucleotide sequence:\")\n",
    "        if user == \"\":\n",
    "            break\n",
    "    \n",
    "        seq_inst = Sequence(user)\n",
    "        inst_list.append(seq_inst)\n",
    "    \n",
    "    return inst_list\n",
    "\n",
    "\n",
    "# Function takes list of Sequence instances, and an integer specifying kmer length. Returns dictionary\n",
    "# with all possible kmers, where the kmer is the key and the number of occurences is the value.\n",
    "def kmerDict(seq_instances, size):\n",
    "    ret_dict = {}\n",
    "    kmer_length = size - 1\n",
    "    \n",
    "    for inst in seq_instances:\n",
    "        if inst.seq_type == \"N\": # only considers nucleotide strings\n",
    "            \n",
    "            for start_pos in range(len(inst.seq) - kmer_length):\n",
    "                substring = inst.seq[start_pos:start_pos+size]\n",
    "                ret_dict.setdefault(substring, 0)\n",
    "                ret_dict[substring] += 1\n",
    "    \n",
    "    return ret_dict          \n",
    "  \n",
    "# creates list of Sequence instances\n",
    "seq_instances = sequenceList()   \n",
    "\n",
    "# passes list as a parameter to kmerDict and returns all possible 3mers\n",
    "three_mer = kmerDict(seq_instances, 3)\n",
    "print(three_mer)\n",
    "\n",
    "# all possible 4mers\n",
    "four_mer = kmerDict(seq_instances, 4)\n",
    "print(four_mer)\n",
    "\n",
    "# all possible 5mers\n",
    "five_mer = kmerDict(seq_instances, 5)\n",
    "print(five_mer)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3rc1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
